src/recovery/model/change.rs

```diff
@@ -0,0 +1,139 @@
+//! Base data model and traits for a "change data capture"-like
+//! system.
+//!
+//! The idea here is to model a stream of keyed changes to a key-value
+//! DB **store**. Streams of changes can be written and read to / from
+//! the store. The store itself does _not_ model time, so changes to
+//! the same key will be lost; only the final state must be
+//! retained. (But it is okay to retain more than just the final
+//! state; it's just not required.)
+
+use serde::Deserialize;
+use serde::Serialize;
+use std::cell::RefCell;
+use std::collections::{BTreeMap, HashMap};
+use std::hash::Hash;
+use std::ops::DerefMut;
+use std::rc::Rc;
+
+/// Represents a change to a value.
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub(crate) enum Change<V> {
+    Upsert(V),
+    Discard,
+}
+
+/// Uses to store a "type of change", but not the change data itself.
+pub(crate) type ChangeType = Change<()>;
+
+impl<V> Change<V> {
+    pub(crate) fn map<U>(self, f: impl Fn(V) -> U) -> Change<U> {
+        match self {
+            Self::Upsert(v) => Change::Upsert(f(v)),
+            Self::Discard => Change::Discard,
+        }
+    }
+
+    pub(crate) fn typ(&self) -> ChangeType {
+        match self {
+            Self::Upsert(..) => Change::Upsert(()),
+            Self::Discard => Change::Discard,
+        }
+    }
+}
+
+/// A "keyed change" representing changing a value for a key in a
+/// store.
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub(crate) struct KChange<K, V>(pub(crate) K, pub(crate) Change<V>);
+
+/// Allows reifying keyed changes into a store.
+///
+/// The order is important. This store must apply the changes in
+/// order.
+pub(crate) trait KWriter<K, V> {
+    fn write(&mut self, kchange: KChange<K, V>);
+
+    fn write_many(&mut self, batch: Vec<KChange<K, V>>) {
+        for change in batch {
+            self.write(change);
+        }
+    }
+}
+
+/// Allows reading keyed changes from a store.
+///
+/// This is not exactly the inverse of [`KWriter`]: a store is allowed
+/// to coallece writes that happen to the same key.
```

Comment:
coalesce?

Reply:
Yeah, this is a tricky concept that I should clarify. The reader and writer contracts do not say that "every write I promise to read back". They say "if you apply all reads to an empty K-V store, you'll get the same store state".

Example:

1. Write `Change::Upsert("a", 1)`, `Change::Upsert("b", 2)`, `Change::Upsert("a", 2)`.
2. That means the resulting store is something like `{"a": 2, "b": 2}`.
3. It's thus valid for a read of that store to result in `Change::Upsert("b", 2)`, `Change::Upsert("a", 2)`, since those are the only writes needed to recreate the same state.

"Coalesce" is a word that sort of gets at that: combining writes to only have the latest ones.

Let me try updating the rustdocs here to be more clear.

Reply:
Nope, I was pointing out a typo `coallece`->`coalesce`

Reply:
Hah. Awkward for me. Well, more documentation doesn't hurt anyway.